cranelift-codegen/meta/src/isa/x86/encodings.rs

```diff
@@ -1669,156 +1575,138 @@ pub(crate) fn define(
         x86_cvtt2si.bind(I32).bind(F64),
-        rec_rfurm.opcodes(vec![0xf2, 0x0f, 0x2c]),
+        rec_rfurm.opcodes(&CVTTSD2SI),
     );
     e.enc64(
         x86_cvtt2si.bind(I64).bind(F64),
-        rec_rfurm.opcodes(vec![0xf2, 0x0f, 0x2c]).rex().w(),
+        rec_rfurm.opcodes(&CVTTSD2SI).rex().w(),
     );
 
     // Exact square roots.
-    e.enc_both(sqrt.bind(F32), rec_furm.opcodes(vec![0xf3, 0x0f, 0x51]));
-    e.enc_both(sqrt.bind(F64), rec_furm.opcodes(vec![0xf2, 0x0f, 0x51]));
+    e.enc_both(sqrt.bind(F32), rec_furm.opcodes(&SQRTSS));
+    e.enc_both(sqrt.bind(F64), rec_furm.opcodes(&SQRTSD));
 
     // Rounding. The recipe looks at the opcode to pick an immediate.
     for inst in &[nearest, floor, ceil, trunc] {
-        e.enc_both_isap(
-            inst.bind(F32),
-            rec_furmi_rnd.opcodes(vec![0x66, 0x0f, 0x3a, 0x0a]),
-            use_sse41,
-        );
-        e.enc_both_isap(
-            inst.bind(F64),
-            rec_furmi_rnd.opcodes(vec![0x66, 0x0f, 0x3a, 0x0b]),
-            use_sse41,
-        );
+        e.enc_both_isap(inst.bind(F32), rec_furmi_rnd.opcodes(&ROUNDSS), use_sse41);
+        e.enc_both_isap(inst.bind(F64), rec_furmi_rnd.opcodes(&ROUNDSD), use_sse41);
     }
 
     // Binary arithmetic ops.
-    for &(inst, opc) in &[
-        (fadd, 0x58),
-        (fsub, 0x5c),
-        (fmul, 0x59),
-        (fdiv, 0x5e),
-        (x86_fmin, 0x5d),
-        (x86_fmax, 0x5f),
-    ] {
-        e.enc_both(inst.bind(F32), rec_fa.opcodes(vec![0xf3, 0x0f, opc]));
-        e.enc_both(inst.bind(F64), rec_fa.opcodes(vec![0xf2, 0x0f, opc]));
-    }
+    e.enc_both(fadd.bind(F32), rec_fa.opcodes(&ADDSS));
+    e.enc_both(fadd.bind(F64), rec_fa.opcodes(&ADDSD));
+
+    e.enc_both(fsub.bind(F32), rec_fa.opcodes(&SUBSS));
+    e.enc_both(fsub.bind(F64), rec_fa.opcodes(&SUBSD));
+
+    e.enc_both(fmul.bind(F32), rec_fa.opcodes(&MULSS));
+    e.enc_both(fmul.bind(F64), rec_fa.opcodes(&MULSD));
+
+    e.enc_both(fdiv.bind(F32), rec_fa.opcodes(&DIVSS));
+    e.enc_both(fdiv.bind(F64), rec_fa.opcodes(&DIVSD));
+
+    e.enc_both(x86_fmin.bind(F32), rec_fa.opcodes(&MINSS));
+    e.enc_both(x86_fmin.bind(F64), rec_fa.opcodes(&MINSD));
+
+    e.enc_both(x86_fmax.bind(F32), rec_fa.opcodes(&MAXSS));
+    e.enc_both(x86_fmax.bind(F64), rec_fa.opcodes(&MAXSD));
 
     // Binary bitwise ops.
-    for &(inst, opc) in &[(band, 0x54), (bor, 0x56), (bxor, 0x57)] {
-        e.enc_both(inst.bind(F32), rec_fa.opcodes(vec![0x0f, opc]));
-        e.enc_both(inst.bind(F64), rec_fa.opcodes(vec![0x0f, opc]));
-    }
+    e.enc_both(band.bind(F32), rec_fa.opcodes(&ANDPS));
+    e.enc_both(band.bind(F64), rec_fa.opcodes(&ANDPS)); // FIXME: ANDPD? Missing 0x66
+
+    e.enc_both(bor.bind(F32), rec_fa.opcodes(&ORPS));
+    e.enc_both(bor.bind(F64), rec_fa.opcodes(&ORPS)); // FIXME: ORPD? Missing 0x66
+
+    e.enc_both(bxor.bind(F32), rec_fa.opcodes(&XORPS));
+    e.enc_both(bxor.bind(F64), rec_fa.opcodes(&XORPS)); // FIXME: XORPD? Missing 0x66
```

Comment:
The 0x66 here is the difference between `andps` and `andpd`, which are semantically equivalent, so skipping the `0x66` is just saving a byte in the encoding. I think the original idea of `andpd` was to let hardware optimize for the case where the data is coming from or going to an instruction using the f64 format, but in practice, no CPUs are known to care about the difference. So I think we can just add a comment here saying something like "Use the "ps" forms of and/or/xor instead of "pd" because CPU's don't treat them differently, and the "ps" form is one byte shorter."